include/core/CCompressUtils.h

```diff
@@ -37,50 +37,135 @@ namespace core {
     bool addString(const std::string& input);
 
-    //! Get compressed representation.  This will fail if the lengthOnly
-    //! constructor argument was set to true.
+    //! Add a vector of trivially copyable types.
+    //!
+    //! \note Multiple calls to this function without finishing
+    //! are equivalent to deflating or inflating the concatenation
+    //! of the vectors passed in the order they are passed.
+    template<typename T>
+    bool addVector(const std::vector<T>& input) {
+        static_assert(std::is_trivially_copyable<T>::value, "Type must be trivially copyable");
+        if (input.empty()) {
+            return true;
+        }
+        if (m_State == E_Finished) {
+            // If the last round of data processing has finished
+            // and we're adding a new vector then we need to reset
+            // the stream so that a new round starts from scratch.
+            this->reset();
+        }
+        return this->processInput(false, input);
+    }
+
+    //! Get transformed representation.
     //!
-    //! \note The compressed representation is a byte array NOT a string,
+    //! \warning This will fail if the lengthOnly constructor argument
+    //!  was set to true.
+    //!
+    //! \note The output representation is a byte array NOT a string,
     //! and hence not printable.
     //!
-    //! If finish==false then retrieve partial compressed state.
-    bool compressedData(bool finish, TByteVec& result);
+    //! If finish==false then retrieve partial state.
+    bool data(bool finish, TByteVec& result);
 
-    //! Get compressed string length.
+    //! Get transformed data length.
     //!
-    //! If finish==false then retrieve partial compressed length.
-    bool compressedLength(bool finish, size_t& length);
+    //! If finish==false then retrieve partial length.
+    bool length(bool finish, size_t& length);
 
-    //! Reset the compressor.  This will happen automatically when adding a
-    //! new string after having finished the previous compression, but
-    //! sometimes, for example when recovering from an error, it may be
-    //! desirable to explicitly reset the compressor state.
+    //! Reset the underlying stream.  This will happen automatically
+    //! when adding a new string after having finished the previous
+    //! round, but sometimes, for example when recovering from an
+    //! error, it may be desirable to explicitly reset the state.
     void reset();
 
 private:
-    bool doCompress(bool finish, const std::string& input);
+    enum EState { E_Unused, E_Active, E_Finished };
+
+private:
+    static const size_t CHUNK_SIZE{4096};
 
 private:
-    enum EState { E_Unused, E_Compressing, E_Finished };
+    //! Get an unsigned character pointer to the address of the start
+    //! of the vector data.
+    template<typename T>
+    static Bytef* bytes(const std::vector<T>& input) {
+        return const_cast<Bytef*>(reinterpret_cast<const Bytef*>(input.data()));
```

Comment:
consider: zlib becomes much nice w.r.t. `const` if you:

```
#define ZLIB_CONST
```

But this requires zlib >= 1.2.8, I am not sure what versions we use on the different platforms.

Reply:
To use this we'd have to build zlib from source on Linux and ship it with our distribution.

CentOS 6 is using version 1.2.3 (patched for security vulnerabilities thankfully) and even CentOS 7 is only on version 1.2.7.  So it will be many years before we can use this flag with the standard OS version of zlib.

Reply:
:+1: ok, probably not worth the hassle. 1.2.8 is already >5 years old, but it seems not old enough ;-)

Reply:
CentOS 6 is basically a hardened version of Fedora 12, and Fedora 12 was released in 2009!

I think `ZLIB_CONST` went into version 1.2.7.1 though, not 1.2.8, so it is available on CentOS 7.  So maybe we can use it in 8.x if CentOS 7 is the oldest supported platform by then.